ntoskrnl/io/iomgr/deviface.c

```diff
@@ -282,185 +282,239 @@ IopSeparateSymbolicLink(
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    DeviceKeyHandle = InstanceKeyHandle = NULL;
 
-    if (GuidKey != NULL)
-        GuidKeyRealP = GuidKey;
-    else
-        GuidKeyRealP = &GuidKeyReal;
+    RtlInitEmptyUnicodeString(&MungedString, NULL, 0);
+    RtlInitEmptyUnicodeString(&ReferenceString, NULL, 0);
 
-    if (DeviceKey != NULL)
-        DeviceKeyRealP = DeviceKey;
-    else
-        DeviceKeyRealP = &DeviceKeyReal;
+    /* Separate it onto the parts */
+    Status = IopSeparateSymbolicLink(SymbolicLinkName,
+                                     NULL,
+                                     &MungedString,
+                                     NULL,
+                                     &ReferenceString,
+                                     NULL,
+                                     NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to separate symbolic link %wZ with status 0x%08lx\n", SymbolicLinkName, Status);
+        goto Quit;
+    }
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    /* Try to open or create device interface keys */
+    if (Create)
+    {
+        Status = IopCreateRegistryKeyEx(&DeviceKeyHandle,
+                                        ClassHandle,
+                                        &MungedString,
+                                        DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
+                                        REG_OPTION_NON_VOLATILE,
+                                        &DeviceKeyDisposition);
+    }
     else
-        InstanceKeyRealP = &InstanceKeyReal;
-
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
-
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    {
+        Status = IopOpenRegistryKeyEx(&DeviceKeyHandle,
+                                      ClassHandle,
+                                      &MungedString,
+                                      DesiredAccess | KEY_ENUMERATE_SUB_KEYS);
+    }
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ\n", &BaseKeyU);
-        goto cleanup;
+        DPRINT1("Failed to create or open %wZ with status 0x%08lx\n", &MungedString, Status);
+        goto Quit;
     }
 
-    StartPosition = wcschr(SymbolicLinkName->Buffer, L'{');
-    EndPosition = wcschr(SymbolicLinkName->Buffer, L'}');
-    if (!StartPosition || !EndPosition || StartPosition > EndPosition)
+    /* Replace the 1st '\' by '#' in reference string */
+    ReferenceString.Buffer[0] = L'#';
```

Comment:
Are we sure here the reference string always starts with `\` ?
Also, can it contain other backslashes, and if so, what about them: should they be kept or be replaced?

Reply:
Yep, the reference string always starts with `\` when it's a part of the symbolic link and from `#` when it's a part of its registry key name. However, reference string is optional part of symbolic link, so in case it does not exist, `\` and `#` are simply used at the end of it, before NULL terminator. So changing the first character to the `#` and then back to `\` should always work correctly.

Reply:
> Also, can it contain other backslashes, and if so, what about them: should they be kept or be replaced?

No, it can't. Usually it starts from `\` or `#` and a some word after that, in case it exists. Otherwise, it's just `\` or `#`, but if it was properly parsed from symbolic link, it shouldn't contain another occurrences of these characters in another places.

Reply:
Maybe having `ASSERT(ReferenceString.Buffer[0] == L'\\' || ReferenceString.Buffer[0] == L'#');` could help make these facts clear in the code.

Reply:
OK, but I think we can (optionally) also have it as a two separate asserts (and at two different places) instead, because it is changed here and restored below, so it looks more logically correct. Although an assert suggested by you shouldn't also not be hit if the symbolic link is initialized correctly and contains a correct character at the first place. So it can be used as well too.

Reply:
Done. Added an assert.

Reply:
Ah no, sorry, my bad. After adding this assert, it is hit during loading other non-audio drivers (when `OpenRegistryHandlesFromSymbolicLink` is called  from `IoSetDeviceInterfaceState`). According to my debugging results, the first character of symbolic link which is tested for the condition is actually `NULL` in that case (for symbolic links where the reference string is missing). It is initialized correctly only below this assert.
So I wonder is it better to either fix the initialization of the reference string in `IoSetDeviceInterfaceState` or just improve the assert here instead by testing for the null condition too? :thinking:

Reply:
If the Reference string is empty (first character is NULL, and since this is from a UNICODE_STRING, its Length parameter should be == 0, I wonder whether you need to do these IopCreate/OpenRegistryKeyEx calls below ... What do you think?

Reply:
No, the open registry calls below are badly required, since we actually need to access and read the data from/write to the instance subkeys. If the reference string is missing, we nevertheless need to create/open the `#` registry key, since the important data (symbolic link string) is stored there. So it needs to be initialized nevertheless. If the reference string is missing, symbolic link should always have `#` and `\` appended at the end (for registry key name and for symbolic link name appropriately). See the ACPI device class for example (the one which has no reference string in its symbolic link, on the system without this PR applied):
![VirtualBox_ReactOS_30_11_2024_22_30_31](https://github.com/user-attachments/assets/8a80f21f-622b-4193-96c4-7f2467f7e83e)
As you can see, `#` registry key exists there, so it should always exist.  It should never have a `NULL` reference string. That's should be at least `#`.
As I remember, it was also previously confirmed by my debugging analysis from the past years.

Reply:
Can you show me a screenshot of that situation on windows?

Reply:
OK, see the screenshot from my XP SP3:
![VirtualBox_Windows XP_30_11_2024_22_44_15](https://github.com/user-attachments/assets/76a5aa69-6ff1-4ad6-93f0-187dff5c1245)
There is exactly the same thing: in case of missing reference string, the `#` key is created anyway.
The only thing is, `SymbolicLink` value has no `\` appended after the GUID part, same as in ReactOS, so it does not need to be appended in case reference string is `NULL`, but it is badly required for registry key nevertheless.

Reply:
Ok then.

Reply:
OK, so the only thing I need to do is improve the assert condition. Everything else is correct here. :slightly_smiling_face:

Reply:
OK, just improved that assert. Now it does not occur in that case. Thank you. :+1: